### ISSUE TITLE
no-extra-parens with return conditions and ternary if

### ISSUE THREAD

**Issue by rijnhard** [eslint/eslint#6513]
**What version of ESLint are you using?** 2.9

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:** Not Required `'no-extra-parens' : [2, 'all']`

**What did you expect to happen?**
None of the overrides for all will allow this to pass:

```javascript
function detectFormat(filePath) {
    const ext = path.extname(filePath);

    // First issue return result of condition
    return (ext.length > 0);

    // Second issue ternary conditions
    let x = (1 != 0) ? 1 : 2;
    return (ext.length > 0 && ext[0] === '.')
        ? ext.substring(1)
        : ext;
}
```
Firstly allowing brackets around a condition in a return makes it obvious the return it the result of a check. Suggest 'returnCondition' override for `all`.
Secondly when using ternary if statements wrapping the check in brackets makes it more legible. 
I'm not sure if configuring trenary brackets should be part of `no-extra-parens`, I think it would be better suited elsewhere and `no-extra-parens` should ignore ternary statements.

**What actually happened? Please include the actual, raw output from ESLint.**
`ESLint: Gratuitous parentheses around expression. (no-extra-parens)`

**Comment by aubergine10** [comment]
If I understand [ASI](http://inimino.org/~inimino/blog/javascript_semicolons) correctly, `return` is a special case.

For example, this:

```js
    return (ext.length > 0 && ext[0] === '.')
        ? ext.substring(1)
        : ext;
```

Becomes this (correct me if I'm wrong):

```js
    return (ext.length > 0 && ext[0] === '.') ; // ASI
        ? ext.substring(1)
        : ext;
```

When surely you should be doing this:

```js
    return ( (ext.length > 0 && ext[0] === '.')
        ? ext.substring(1)
        : ext
    );
```

**Comment by platinumazure** [comment]
@aubergine10 ASI only occurs when the next statement could stand on its own. It can't occur if the next line is beginning with a binary operator.

Return is special-cased, but only in terms of when return is by itself:

```js
return // ASI will happen here
{};
```

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
